src/main/jni/tiff/doc/man-prebuilt/tiffdither.1
'\" t
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "TIFFDITHER" "1" "Jun 27, 2026" "4.7" "LibTIFF"
.SH NAME
tiffdither \- convert a greyscale TIFF image to bilevel using dithering
.SH SYNOPSIS
.sp
\fBtiffdither\fP [ \fIoptions\fP ] \fIinput.tif\fP \fIoutput.tif\fP
.SH DESCRIPTION
.sp
\fBtiffdither\fP converts a single channel 8\-bit greyscale image to a bilevel image
using Floyd\-Steinberg error propagation with thresholding.
.SH OPTIONS
.INDENT 0.0
.TP
.B \-c compress
Specify the compression to use for data written to the output file:
.TS
box center;
l|l.
T{
Compression
T}	T{
Description
T}
_
T{
none
T}	T{
no compression
T}
_
T{
packbits
T}	T{
PackBits compression
T}
_
T{
lzw
T}	T{
Lempel\-Ziv & Welch compression
T}
_
T{
zip
T}	T{
Deflate compression
T}
_
T{
g3
T}	T{
CCITT Group 3 (T.4) compression
T}
_
T{
g4
T}	T{
for CCITT Group 4 (T.6) compression
T}
.TE
.sp
By default \fBtiffdither\fP
will compress data according to the value of the
\fBCompression\fP tag found in the source file.
.sp
The CCITT Group 3 and Group 4 compression algorithms can only
be used with bilevel data.
.sp
Group 3 compression can be specified together with several
T.4\-specific options:
\fB1d\fP for 1\-dimensional encoding,
\fB2d\fP for 2\-dimensional encoding, and
\fBfill\fP to force each encoded scanline to be zero\-filled so that the
terminating EOL code lies on a byte boundary.
Group 3\-specific options are specified by appending a \fB:\fP\-separated
list to the \fBg3\fP option; e.g. \fB\-c g3:2d:fill\fP
to get 2D\-encoded data with byte\-aligned EOL codes.
.sp
LZW compression can be specified together with a \fBpredictor\fP value.
A predictor value of 2 causes each scanline of the output image to undergo
horizontal differencing before it is encoded; a value
of 1 forces each scanline to be encoded without differencing.
LZW\-specific options are specified by appending a \fB:\fP\-separated
list to the \fBlzw\fP option; e.g. \fB\-c lzw:2\fP for LZW
compression with horizontal differencing.
.UNINDENT
.INDENT 0.0
.TP
.B \-f
Specify the bit fill order to use in writing output data.
By default, \fBtiffdither\fP
will create a new file with the same fill order as the original.
Specifying
\fB\-f lsb2msb\fP will force data to be written with the
\fBFillOrder\fP tag set to \fBLSB2MSB\fP, while
\fB\-f msb2lsb\fP will force data to be written with the
\fBFillOrder\fP tag set to \fBMSB2LSB\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-r
Make each strip have no more than the given number of rows.
.UNINDENT
.INDENT 0.0
.TP
.B \-t
Set the threshold value for dithering.
By default the threshold value is 128.
.UNINDENT
.SH NOTES
.sp
The dither algorithm is taken from the \fBtiffmedian\fP
program (written by Paul Heckbert).
.SH SEE ALSO
.sp
\fI\%pal2rgb\fP (1),
\fI\%fax2tiff\fP (1),
\fI\%tiffinfo\fP (1),
\fI\%tiffcp\fP (1),
\fI\%tiff2bw\fP (3tiff)
.SH AUTHOR
LibTIFF contributors
.SH COPYRIGHT
1988-2025, LibTIFF contributors
.\" Generated by docutils manpage writer.
.


CHANGELOG.txt
0.9.9.3
- Fixed bugs in native and java code

0.9.9.2
- Added support for 16KB Pages. Increase minimum support level to 21. Added riscv64 ABI

0.9.9.1
- Fixed problem with owned file descriptor (https://github.com/Beyka/Android-TiffBitmapFactory/issues/54)

0.9.9.0
- Added support for decoding and converting files from file descriptors

0.9.8.7
- Fixed converting from JPG to TIFF with CCITTRLE, CCITTFAX3, CCITTFAX4 compression schemes

0.9.8.6
- Changed binary files (rebuild)

0.9.8.5
- Added signal handling of SIGSEGV signal before open tiff images to prevent many silent crashes from libtiff

0.9.8.4
- Added possibility to cancel background processing by interrupting of worker thread as usual in Java

0.9.8.3
-Fixed impossibility to convert full image without bounds

0.9.8.2
-Added handling of SIGSEGV crashes from libtiff

0.9.8
-Added support for dirrect converting files to TIFF and from TIFF
-Fixed bug causing crashing while saving tiff

0.9.7.6
-Added support for CCITT modified Huffman RLE (CCITTRLE) compression scheme

0.9.7.5
- Returned previous version of write bitmap to tiff algorithm with some improvements. This changes is due to an error of writing tiff data with bottom orientations

0.9.7.4
- Add converter utils
- Improved write and read algorithms

0.9.7.3
- Add possibility to decode part of image by specify decode area

0.9.7.2
- Changed saving algorithm to use android native bitmap
- Add progress listener for decoding process

0.9.7.1
- Added reading of various tiff tags
- Changed enums

0.9.7
- Added possibility to stop decoding that runs in separate thread
- Fixed crashing when decoding images with bad metadata

0.9.6.2
- Added support for TIFFTAG_XRESOLUTION, TIFFTAG_YRESOLUTION, TIFFTAG_RESOLUTIONUNIT

0.9.6
- Added support for CCITT Group 3 and CCITT Group 4 compression schemes

0.9.5.1
- Fixed decoding algorithm for stripped images

0.9.5 
- Added algorithms for efficient memory usage
- Added switchers for throwing exceptions and using of tiff tag orientation



src/main/jni/tiff/tools/tiffdither.c
/*
 * Copyright (c) 1988-1997 Sam Leffler
 * Copyright (c) 1991-1997 Silicon Graphics, Inc.
 *
 * Permission to use, copy, modify, distribute, and sell this software and
 * its documentation for any purpose is hereby granted without fee, provided
 * that (i) the above copyright notices and this permission notice appear in
 * all copies of the software and related documentation, and (ii) the names of
 * Sam Leffler and Silicon Graphics may not be used in any advertising or
 * publicity relating to the software without the specific, prior written
 * permission of Sam Leffler and Silicon Graphics.
 *
 * THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY
 * WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
 *
 * IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
 * ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
 * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF
 * LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE
 * OF THIS SOFTWARE.
 */

#include "libport.h"
#include "tif_config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#include "tiffio.h"
#include "tiffiop.h"

#ifndef EXIT_SUCCESS
#define EXIT_SUCCESS 0
#endif
#ifndef EXIT_FAILURE
#define EXIT_FAILURE 1
#endif

#define streq(a, b) (strcmp(a, b) == 0)
#define strneq(a, b, n) (strncmp(a, b, n) == 0)

#define CopyField(tag, v)                                                      \
    if (TIFFGetField(in, tag, &v))                                             \
    TIFFSetField(out, tag, v)
#define CopyFieldFloat(tag, v)                                                 \
    if (TIFFGetField(in, tag, &v))                                             \
    TIFFSetField(out, tag, (double)(v))

uint32_t imagewidth;
uint32_t imagelength;
int threshold = 128;

static void usage(int code);

/*
 * Floyd-Steinberg error propragation with threshold.
 * This code is stolen from tiffmedian.
 */
static int fsdither(TIFF *in, TIFF *out)
{
    unsigned char *outline, *inputline, *inptr;
    short *thisline, *nextline, *tmpptr;
    unsigned char *outptr;
    short *thisptr, *nextptr;
    uint32_t i, j;
    uint32_t imax, jmax;
    int lastline, lastpixel;
    int bit;
    tsize_t outlinesize;
    int errcode = 0;

    imax = imagelength - 1;
    jmax = imagewidth - 1;
    inputline = (unsigned char *)_TIFFmalloc(TIFFScanlineSize(in));
    thisline = (short *)_TIFFmalloc(
        TIFFSafeMultiply(tmsize_t, imagewidth, sizeof(short)));
    nextline = (short *)_TIFFmalloc(
        TIFFSafeMultiply(tmsize_t, imagewidth, sizeof(short)));
    outlinesize = TIFFScanlineSize(out);
    outline = (unsigned char *)_TIFFmalloc(outlinesize);
    if (!(inputline && thisline && nextline && outline))
    {
        fprintf(stderr, "Out of memory.\n");
        goto skip_on_error;
    }
    if (imagewidth > TIFFScanlineSize(in))
    {
        fprintf(stderr, "Image width exceeds scanline size.\n");
        goto skip_on_error;
    }

    /*
     * Get first line
     */
    if (TIFFReadScanline(in, inputline, 0, 0) <= 0)
        goto skip_on_error;

    inptr = inputline;
    nextptr = nextline;
    for (j = 0; j < imagewidth; ++j)
        *nextptr++ = *inptr++;
    for (i = 0; i < imagelength; ++i)
    {
        tmpptr = thisline;
        thisline = nextline;
        nextline = tmpptr;
        lastline = (i == imax);
        if (TIFFReadScanline(in, inputline, i, 0) <= 0)
            goto skip_on_error;
        inptr = inputline;
        nextptr = nextline;
        for (j = 0; j < imagewidth; ++j)
            *nextptr++ = *inptr++;
        thisptr = thisline;
        nextptr = nextline;
        _TIFFmemset(outptr = outline, 0, outlinesize);
        bit = 0x80;
        for (j = 0; j < imagewidth; ++j)
        {
            int v;

            lastpixel = (j == jmax);
            v = *thisptr++;
            if (v < 0)
                v = 0;
            else if (v > 255)
                v = 255;
            if (v > threshold)
            {
                *outptr |= (unsigned char)bit;
                v -= 255;
            }
            bit >>= 1;
            if (bit == 0)
            {
                outptr++;
                bit = 0x80;
            }
            if (!lastpixel)
                thisptr[0] = (short)(thisptr[0] + v * 7 / 16);
            if (!lastline)
            {
                if (j != 0)
                    nextptr[-1] = (short)(nextptr[-1] + v * 3 / 16);
                *nextptr = (short)(*nextptr + v * 5 / 16);
                nextptr++;
                if (!lastpixel)
                    nextptr[0] = (short)(nextptr[0] + v / 16);
            }
        }
        if (TIFFWriteScanline(out, outline, i, 0) < 0)
            goto skip_on_error;
    }
    goto exit_label;

skip_on_error:
    errcode = 1;
exit_label:
    _TIFFfree(inputline);
    _TIFFfree(thisline);
    _TIFFfree(nextline);
    _TIFFfree(outline);
    return errcode;
}

static uint16_t compression = COMPRESSION_PACKBITS;
static uint16_t predictor = 0;
static uint32_t group3options = 0;

static void processG3Options(char *cp)
{
    if ((cp = strchr(cp, ':')))
    {
        do
        {
            cp++;
            if (strneq(cp, "1d", 2))
                group3options &= ~(uint32_t)GROUP3OPT_2DENCODING;
            else if (strneq(cp, "2d", 2))
                group3options |= GROUP3OPT_2DENCODING;
            else if (strneq(cp, "fill", 4))
                group3options |= GROUP3OPT_FILLBITS;
            else
                usage(EXIT_FAILURE);
        } while ((cp = strchr(cp, ':')));
    }
}

static int processCompressOptions(char *opt)
{
    if (streq(opt, "none"))
        compression = COMPRESSION_NONE;
    else if (streq(opt, "packbits"))
        compression = COMPRESSION_PACKBITS;
    else if (strneq(opt, "g3", 2))
    {
        processG3Options(opt);
        compression = COMPRESSION_CCITTFAX3;
    }
    else if (streq(opt, "g4"))
        compression = COMPRESSION_CCITTFAX4;
    else if (strneq(opt, "lzw", 3))
    {
        char *cp = strchr(opt, ':');
        if (cp)
            predictor = (uint16_t)atoi(cp + 1);
        compression = COMPRESSION_LZW;
    }
    else if (strneq(opt, "zip", 3))
    {
        char *cp = strchr(opt, ':');
        if (cp)
            predictor = (uint16_t)atoi(cp + 1);
        compression = COMPRESSION_ADOBE_DEFLATE;
    }
    else
        return (0);
    return (1);
}

int main(int argc, char *argv[])
{
    TIFF *in, *out;
    uint16_t samplesperpixel, bitspersample = 1, shortv;
    float floatv;
    char thing[1024];
    uint32_t rowsperstrip = (uint32_t)-1;
    uint16_t fillorder = 0;
    int c;
#if !HAVE_DECL_OPTARG
    extern int optind;
    extern char *optarg;
#endif

    while ((c = getopt(argc, argv, "c:f:r:t:h")) != -1)
        switch (c)
        {
            case 'c': /* compression scheme */
                if (!processCompressOptions(optarg))
                    usage(EXIT_FAILURE);
                break;
            case 'f': /* fill order */
                if (streq(optarg, "lsb2msb"))
                    fillorder = FILLORDER_LSB2MSB;
                else if (streq(optarg, "msb2lsb"))
                    fillorder = FILLORDER_MSB2LSB;
                else
                    usage(EXIT_FAILURE);
                break;
            case 'r': /* rows/strip */
                rowsperstrip = (uint32_t)atoi(optarg);
                break;
            case 't':
                threshold = atoi(optarg);
                if (threshold < 0)
                    threshold = 0;
                else if (threshold > 255)
                    threshold = 255;
                break;
            case 'h':
                usage(EXIT_SUCCESS);
                /*NOTREACHED*/
                break;
            case '?':
                usage(EXIT_FAILURE);
                /*NOTREACHED*/
                break;
            default:
                break;
        }
    if (argc - optind < 2)
        usage(EXIT_FAILURE);
    in = TIFFOpen(argv[optind], "r");
    if (in == NULL)
        return (EXIT_FAILURE);
    TIFFGetField(in, TIFFTAG_SAMPLESPERPIXEL, &samplesperpixel);
    if (samplesperpixel != 1)
    {
        fprintf(stderr, "%s: Not a b&w image.\n", argv[0]);
        TIFFClose(in);
        return (EXIT_FAILURE);
    }
    TIFFGetField(in, TIFFTAG_BITSPERSAMPLE, &bitspersample);
    if (bitspersample != 8)
    {
        fprintf(stderr, " %s: Sorry, only handle 8-bit samples.\n", argv[0]);
        TIFFClose(in);
        return (EXIT_FAILURE);
    }
    out = TIFFOpen(argv[optind + 1], "w");
    if (out == NULL)
    {
        TIFFClose(in);
        return (EXIT_FAILURE);
    }

    do
    {
        CopyField(TIFFTAG_IMAGEWIDTH, imagewidth);
        TIFFGetField(in, TIFFTAG_IMAGELENGTH, &imagelength);
        TIFFSetField(out, TIFFTAG_IMAGELENGTH, imagelength - 1);
        TIFFSetField(out, TIFFTAG_BITSPERSAMPLE, 1);
        TIFFSetField(out, TIFFTAG_SAMPLESPERPIXEL, 1);
        TIFFSetField(out, TIFFTAG_PLANARCONFIG, PLANARCONFIG_CONTIG);
        TIFFSetField(out, TIFFTAG_COMPRESSION, compression);
        if (fillorder)
            TIFFSetField(out, TIFFTAG_FILLORDER, fillorder);
        else
            CopyField(TIFFTAG_FILLORDER, shortv);
        snprintf(thing, sizeof(thing), "Dithered B&W version of %s",
                 argv[optind]);
        TIFFSetField(out, TIFFTAG_IMAGEDESCRIPTION, thing);
        CopyField(TIFFTAG_PHOTOMETRIC, shortv);
        CopyField(TIFFTAG_ORIENTATION, shortv);
        CopyFieldFloat(TIFFTAG_XRESOLUTION, floatv);
        CopyFieldFloat(TIFFTAG_YRESOLUTION, floatv);
        CopyField(TIFFTAG_RESOLUTIONUNIT, shortv);
        rowsperstrip = TIFFDefaultStripSize(out, rowsperstrip);
        TIFFSetField(out, TIFFTAG_ROWSPERSTRIP, rowsperstrip);
        switch (compression)
        {
            case COMPRESSION_CCITTFAX3:
                TIFFSetField(out, TIFFTAG_GROUP3OPTIONS, group3options);
                break;
            case COMPRESSION_LZW:
            case COMPRESSION_ADOBE_DEFLATE:
            case COMPRESSION_DEFLATE:
                if (predictor)
                    TIFFSetField(out, TIFFTAG_PREDICTOR, predictor);
                break;
            default:
                break;
        }
        fsdither(in, out);
        TIFFWriteDirectory(out);
    } while (TIFFReadDirectory(in));

    TIFFClose(in);
    TIFFClose(out);
    return (EXIT_SUCCESS);
}

static const char usage_info[] =
    "Convert a greyscale image to bilevel using dithering\n\n"
    "usage: tiffdither [options] input.tif output.tif\n"
    "where options are:\n"
    " -r #      make each strip have no more than # rows\n"
    " -t #      set the threshold value for dithering (default 128)\n"
    " -f lsb2msb    force lsb-to-msb FillOrder for output\n"
    " -f msb2lsb    force msb-to-lsb FillOrder for output\n"
    "\n"
#ifdef LZW_SUPPORT
    " -c lzw[:opts] compress output with Lempel-Ziv & Welch encoding\n"
    "    #          set predictor value\n"
    "    For example, -c lzw:2 for LZW-encoded data with horizontal "
    "differencing\n"
#endif
#ifdef ZIP_SUPPORT
    " -c zip[:opts] compress output with deflate encoding\n"
    "    #          set predictor value\n"
#endif
#ifdef PACKBITS_SUPPORT
    " -c packbits   compress output with packbits encoding\n"
#endif
#ifdef CCITT_SUPPORT
    " -c g3[:opts]  compress output with CCITT Group 3 encoding\n"
    "    Group 3 options:\n"
    "    1d         use default CCITT Group 3 1D-encoding\n"
    "    2d         use optional CCITT Group 3 2D-encoding\n"
    "    fill       byte-align EOL codes\n"
    "    For example, -c g3:2d:fill for G3-2D-encoded data with byte-aligned "
    "EOLs\n"
    " -c g4         compress output with CCITT Group 4 encoding\n"
#endif
#if defined(LZW_SUPPORT) || defined(ZIP_SUPPORT) ||                            \
    defined(PACKBITS_SUPPORT) || defined(CCITT_SUPPORT)
    " -c none       use no compression algorithm on output\n"
#endif
    "\n";

static void usage(int code)
{
    FILE *out = (code == EXIT_SUCCESS) ? stdout : stderr;

    fprintf(out, "%s\n\n", TIFFGetVersion());
    fprintf(out, "%s", usage_info);
    exit(code);
}
